File: ruby/test/wire/issue_clusters_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class IssueClustersWireTest < WireMockTestCase
  def setup
    super

    @client = Sikaru::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_issue_clusters_list_issue_clusters_with_wiremock
    test_id = "issue_clusters.list_issue_clusters.0"

    @client.issue_clusters.list_issue_clusters(
      project_id: "project_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "issue_clusters.list_issue_clusters.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/issue-clusters",
      query_params: nil,
      expected: 1
    )
  end

  def test_issue_clusters_upsert_issue_cluster_with_wiremock
    test_id = "issue_clusters.upsert_issue_cluster.0"

    @client.issue_clusters.upsert_issue_cluster(
      project_id: "project_id",
      label: "label",
      trace_ids: ["traceIds"],
      request_options: {
        additional_headers: {
          "X-Test-Id" => "issue_clusters.upsert_issue_cluster.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/issue-clusters",
      query_params: nil,
      expected: 1
    )
  end

  def test_issue_clusters_mine_project_issue_clusters_with_wiremock
    test_id = "issue_clusters.mine_project_issue_clusters.0"

    @client.issue_clusters.mine_project_issue_clusters(
      project_id: "project_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "issue_clusters.mine_project_issue_clusters.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/issue-clusters/mine",
      query_params: nil,
      expected: 1
    )
  end

  def test_issue_clusters_get_issue_cluster_with_wiremock
    test_id = "issue_clusters.get_issue_cluster.0"

    @client.issue_clusters.get_issue_cluster(
      project_id: "project_id",
      cluster_id: "cluster_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "issue_clusters.get_issue_cluster.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/issue-clusters/cluster_id",
      query_params: nil,
      expected: 1
    )
  end

  def test_issue_clusters_update_issue_cluster_status_with_wiremock
    test_id = "issue_clusters.update_issue_cluster_status.0"

    @client.issue_clusters.update_issue_cluster_status(
      project_id: "project_id",
      cluster_id: "cluster_id",
      status: "status",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "issue_clusters.update_issue_cluster_status.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "PATCH",
      url_path: "/v1/projects/project_id/issue-clusters/cluster_id",
      query_params: nil,
      expected: 1
    )
  end

  def test_issue_clusters_propose_issue_cluster_fix_with_wiremock
    test_id = "issue_clusters.propose_issue_cluster_fix.0"

    @client.issue_clusters.propose_issue_cluster_fix(
      project_id: "project_id",
      cluster_id: "cluster_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "issue_clusters.propose_issue_cluster_fix.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/issue-clusters/cluster_id/propose-fix",
      query_params: nil,
      expected: 1
    )
  end
end

File: rust/tests/issue_clusters_test.rs
use sikaru::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_issue_clusters_list_issue_clusters_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .issue_clusters
        .list_issue_clusters(
            &"project_id".to_string(),
            &ListIssueClustersQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/projects/project_id/issue-clusters", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_issue_clusters_upsert_issue_cluster_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .issue_clusters
        .upsert_issue_cluster(
            &"project_id".to_string(),
            &UpsertIssueClusterRequest {
                label: "label".to_string(),
                trace_ids: vec!["traceIds".to_string()],
                agent_id: None,
                impact: None,
                issue_id: None,
                issue_title: None,
                signal_tags: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/projects/project_id/issue-clusters",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_issue_clusters_mine_project_issue_clusters_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .issue_clusters
        .mine_project_issue_clusters(
            &"project_id".to_string(),
            &MineProjectIssueClustersQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/projects/project_id/issue-clusters/mine",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_issue_clusters_get_issue_cluster_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .issue_clusters
        .get_issue_cluster(&"project_id".to_string(), &"cluster_id".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/v1/projects/project_id/issue-clusters/cluster_id",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_issue_clusters_update_issue_cluster_status_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .issue_clusters
        .update_issue_cluster_status(
            &"project_id".to_string(),
            &"cluster_id".to_string(),
            &UpdateIssueClusterStatusRequest {
                status: "status".to_string(),
                reason: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "PATCH",
        "/v1/projects/project_id/issue-clusters/cluster_id",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_issue_clusters_propose_issue_cluster_fix_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .issue_clusters
        .propose_issue_cluster_fix(&"project_id".to_string(), &"cluster_id".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/projects/project_id/issue-clusters/cluster_id/propose-fix",
        None,
        1,
    )
    .await
    .unwrap();
}
